narwhal/primary/src/state_handler.rs

```diff
@@ -70,13 +226,37 @@ impl StateHandler {
         if let Some(sender) = &self.tx_committed_own_headers {
             let _ = sender.send((commit_round, own_rounds_committed)).await;
         }
+
+        // Process committed system messages.
+        if let Some(randomness_state) = self.randomness_state.as_mut() {
+            for certificate in certificates {
+                let header = certificate.header();
+                for message in header.system_messages() {
+                    match message {
+                        SystemMessage::DkgMessage(msg) => randomness_state.add_message(msg.clone()),
+                        SystemMessage::DkgConfirmation(conf) => {
+                            randomness_state.add_confirmation(conf.clone())
+                        }
+                    }
+                }
+            }
+            // Once all messages in the new commit are saved, advance the random
+            // beacon protocol if possible.
+            randomness_state.advance(&self.tx_system_messages).await;
```

Comment:
I want to read the proposal/protocol more in depth, but are there any requirements at this point on crash recovery? Or we don't have very strict requirements here? (ex worst case scenario we'll just skip some rounds without any randomness created until the next ones)

Reply:
Yeah as you said nothing should be expected to be replayed via consensus to state handler so for sure things can be missed on restart. Sounds good addressing that on separate PR.